### ISSUE TITLE
No constructor found

### ISSUE THREAD

**Issue by JoshuaKolden** [qmlweb/qmlweb#277]
I'm getting strange results trying to get a basic test working.

I've create a new node project.  I've installed `gulp-qmlweb`.  

I've greated a gulp taks to copy qml from src to dist.

```javascript
gulp.task('qml', function() {
  gulp.src('src/qml/**/*.qml').pipe(gulp.dest('dist/qml'))
  return gulp.src(['src/qml/**/*.qml', 'src/qml/**/*.js', 'src/qml/**/qmldir'])
    .pipe(qml())
    .pipe(concat('qrc.js'))
    .pipe(gulp.dest('src/js'));
});
```

I find that I have to also copy the QML files to the dist even though I would think they would work via the qrc.js.

I have the following simple index.html.

```html
<html>
<head>
  <title>Test qmlweb</title>

  <!--build:css css/styles.min.css-->
  <link rel="stylesheet" href="css/styles.css">
  <!--endbuild-->
  <script src="https://ajax.googleapis.com/ajax/libs/jquery/1.12.2/jquery.min.js"></script>

</head>
  <body style="margin: 0;" data-qml="src/qml/main.qml">
    If you see this, main.qml is not working.

    <!--build:js js/main.min.js -->>
    <script type="text/javascript" src="js/qrc.js"></script>
    <script type="text/javascript" src="js/main.js"></script>
    <!-- endbuild -->
    <script type="text/javascript" src="lib/qt.js"></script>
  </body>
</html>
```

I have to copy lib/* from qmlweb manually into my dest folder, so I'm unclear on what is the purpose of `npm install qmlweb`.

With these steps I have main.qml and any other custom qml files rendering.  Thought I don't know if they are being rendered dynamically or if it's using the qrc.js file.

However I can't use `QtQuick.Controls`.  If I put a `TextField` in my main.qml I get an error in the browser console: 

```
[Error] Error: No constructor found for TextField
  construct (qt.js:689:83)
  QMLList (qt.js:1540)
  set (qt.js:1699)
  applyProperties (qt.js:802)
  construct (qt.js:702)
  QMLList (qt.js:1540)
  set (qt.js:1699)
  applyProperties (qt.js:802)
  construct (qt.js:702)
  QMLList (qt.js:1541)
  set (qt.js:1699)
  applyProperties (qt.js:802)
  construct (qt.js:702)
  QMLList (qt.js:1540)
  set (qt.js:1699)
  applyProperties (qt.js:802)
  construct (qt.js:702)
  createObject (qt.js:2576)
  loadQMLTree (qt.js:1094)
  loadFile (qt.js:1065)
  (anonymous function) (qt.js:936)
```

Note I'm using `browser-sync` to serve my ./dist folder for testing if that's relevant.

Any ideas?

**Comment by JoshuaKolden** [comment]
Also I notice something that strikes me as odd in the qt.js file.  Sometimes the constructor is defined as a named function inline, and sometimes it just refers to a function name.  

```javascript
registerQmlType({
    module: 'QtQuick.Controls',
    name: 'TextField',
    versions: /.*/,
    baseClass: 'QtQuick.Item',
    constructor: QMLTextInput
});

function QMLTextInput(meta) {
    callSuper(this, meta);
...
}
```

vs:

```javascript
registerQmlType({
    module: 'QtQuick',
    name: 'ListElement',
    versions: /.*/,
    baseClass: 'QtQml.QtObject',
    constructor: function QMLListElement(meta) {
        callSuper(this, meta);

        for (var i in meta.object) {
            if (i[0] != "$") {
                createProperty("variant", this, i);
            }
        }
        applyProperties(meta.object, this, this, this.$context);
    }
});
```

**Comment by stephenmdangelo** [comment]
Check if my import fixes in #239 help you: https://github.com/qmlweb/qmlweb/pull/239

Without it, you generally have to include all of the same "import" lines in all of your QML files, otherwise the import statements get mixed up with each other. 

I've only been working with non-gulp-compiled QML files, though, so YMMV. 

In regards to registerQmlType, there's currently a move to a nicer syntax as you can see with the Animation and Connections components, everything just hasn't been ported over yet.

**Comment by ChALkeR** [comment]
> Sometimes the constructor is defined as a named function inline, and sometimes it just refers to a function name.

#271 (WIP) is going to fix that.

**Comment by ChALkeR** [comment]
@JoshuaKolden Could you check if #271 fixes your issue? I can't reproduce this without a qml testcase, btw.

**Comment by JoshuaKolden** [comment]
Nope.  The line numbers are a little different in the stack trace, but otherwise it's basically the same error.

```javascript
[Error] Error: No constructor found for TextField
  construct (qt.js:701:83)
  QMLList (qt.js:1552)
  set (qt.js:1710)
  applyProperties (qt.js:814)
  construct (qt.js:714)
  QMLList (qt.js:1552)
  set (qt.js:1710)
  applyProperties (qt.js:814)
  construct (qt.js:714)
  QMLList (qt.js:1552)
  set (qt.js:1710)
  applyProperties (qt.js:814)
  construct (qt.js:714)
  QMLList (qt.js:1553)
  set (qt.js:1710)
  applyProperties (qt.js:814)
  construct (qt.js:714)
  QMLList (qt.js:1552)
  set (qt.js:1710)
  applyProperties (qt.js:814)
  construct (qt.js:714)
  createObject (qt.js:2622)
  loadQMLTree (qt.js:1106)
  loadFile (qt.js:1077)
  (anonymous function) (qt.js:948)
```

**Comment by JoshuaKolden** [comment]
It looks like it's trying to pull TextField from the server which of course does't have it,

How does anyone use this?  I bet I've just got it setup wrong, but I can't debug that without some setup instructions.  The Readme.md isn't even self-consistent.

**Comment by Plaristote** [comment]
That's the expected behavior : it means TextField doesn't exist from the context in which you ask for it. So it tries to find out whether TextField is a type that you defined yourself, by trying to download an hypothetical qml file.

In the beginnings, loaded types were accessible from any context (the issue being that if you important QtQuick.Controls in any file, it would be important in *all* the files). I'm not sure if we changed that or not already, but if we did, maybe something's wrong there.

Otherwise, this is basically what would happen if you didn't add QtQuick.Controls to your import.

**Comment by stephenmdangelo** [comment]
@Plaristote 

> In the beginnings, loaded types were accessible from any context (the issue being that if you imported QtQuick.Controls in any file, it would be important in all the files). I'm not sure if we changed that or not already, but if we did, maybe something's wrong there.

It appears that now imports in various contexts can overwrite each other. See the test case added in #239.

@JoshuaKolden did you try #239 or including the same import statements in every QML file you are using?

**Comment by JoshuaKolden** [comment]
@Plaristote I do not think that word means what you think it means.

"That's the expected behavior...",  I'm sorry, but no.  This is the very definition of unexpected behavior.  When a user writes:

```QML
import QtQuick 2.0
import QtQuick.Controls 1.4

Rectangle { id: main
  anchors.fill: parent
  color: "#1B2431"
  
  // ...

  TextField { id: filterInput
    anchors.fill: parent
    placeholderText: "Filter"
  }
}
```

The system responding with "WTF is a TextField" (in effect) is absolutely not the expected behavior!  TextField is defined IN THE qt.js file for **crying out loud** (line #3527, it the above PR if you're curious).  Never mind the copious qt.io documentation.

Nope.  You can't bring red balloons to the party, and then tell the kids that wanting a red balloon is unexpected behavior.

I want my red balloon.  Just say'n

**Comment by JoshuaKolden** [comment]
@stephenmdangelo No I have not tried #239 / tried to duplicate the imports yet.  Will do, and report back.  Thx.

**Comment by ChALkeR** [comment]
@JoshuaKolden The provided example in https://github.com/qmlweb/qmlweb/issues/277#issuecomment-228530037 works for me (apart from `placeholderText` not being supported atm).

Could you provide a minimal reproducible example that fails? Perhaps uploading it somewhere will help.

**Comment by Plaristote** [comment]
@JoshuaKolden Well, how about you start using QtCreator, and try to see what happens when you use `TextField` without importing `QtQuick.Contrlols`. The behavior that I expect is that it won't work.

You can expect whatever you want, we're not the ones making the specs. We're not going to make a type available without the import being performed, since this is not how the original QML does thing.

Now, did you or did you not forget to import QtQuick.Controls, and if you didn't, can you provide us with an example that reproduces the issue ?

**Comment by JoshuaKolden** [comment]
@ChALkeR Will do. I'll check @stephenmdangelo's #239, and if that doesn't address it then I'll create a minimal project example of what I'm doing.

@Plaristote Sorry, who uses QtCreator?  I think I was pretty clear that my (hopefully humorous?) 'unexpected behavior' is pursuant to the expectation that  `import QtQuick.Controls 1.4` should work. 

Poking fun aside, as I said at the top I'm not even clear that I'm following the proper steps to add qmlweb to a trivial a web project.  If I can sort that workflow out, then I can help with documentation, PRs and antagonistic support.

**Comment by Plaristote** [comment]
@JoshuaKolden I followed a few step to check if I could use `TextField` in an example built from scratch.

For reference, you could try following these steps, and see if it reacts differently from what you've done before:
```
$> git clone  https://github.com/qmlweb/qmlweb.git
$> cd qmlweb
$> npm install
$> npm run test # or build, but test will also build and gives me reassurance about everything working fine
$> touch index.html main.qml
$> npm install http-server
$> http-server # will launch a client bound to port 8080
```

As for the index.html:
```
<html>
  <head>
    <title>QML Example</title>
    <script type="text/javascript" src="/lib/qt.js"></script>
  </head>
  <body style="margin:0" data-qml="main.qml">
  </body>
</html>
```

And for the main.qml file:
```
import QtQuick 2.0
import QtQuick.Controls 1.4

Rectangle {
  TextField {
  }
}
```

Though whatever you might've done that might've broken that workflow, it would still be useful for us to see what happened.

**Comment by JoshuaKolden** [comment]
@Plaristote, great. What I did was follow the instructions in the Readme.md.  Your process assume I'm modifying qmlweb itself, which (of course) no one would be doing in normal use.

But give me a sec.  Let me give that a go.

Here's the deal though.  If it works I'll expect you to post those instructions to the Readme.md, and if it doesn't I'll be sending you a red balloon for your birthday for longer than you can comfortably explain to your wife. ;)

**Comment by JoshuaKolden** [comment]
Ok the results are kind of a toss up.  Your steps, as written, don't work.  

```bash
$> npm run build
```
Is not optional.

However, the `Readme.md` as it currently stands does have that requirement, though the example HTML is different. 

My verdict is you have to update the Readme.md, AND red balloon for your birthday.  Boy this really is going to be awkward.  Um, so when is your birthday?

**Comment by ChALkeR** [comment]
@JoshuaKolden `npm run build` is optional and it should not be required, unless you try to run from `master` or other non-release version. Any pull requests to improve the Readme are welcome, though.
I still haven't seen any exact steps to reproduce a problem here.

**Comment by Plaristote** [comment]
@Chalker Actually, that's not even the problem here.

Apparently, `npm run test` didn't run the build task with him. Which I find weird, because I personally never run `npm run build`, I always only run `npm run test`, and it always did the trick for me.

This only looks weirder and weirder. Since we've both re-cloned the repository today, I can only assume that the node dependencies we have do use the same versions. Could it be possibly linked to different version on nodejs ? I'm using version 6.2.2 right now.

**Comment by ChALkeR** [comment]
@Plaristote `npm run test` does not depend on the `build` task, and it does not produce the `./lib` directory anymore. If one is running anything that is not release, he or she should run `npm run build` to get files in the `./lib` directory.

**Comment by JoshuaKolden** [comment]
So, we are in agreement, the Readme.md is wrong and the above steps require `npm run build`, yes?

In addition to this, returning to my first point, how is someone supposed to 'use' this in their own project?  How, for example, does someone create a QML based web site?  The workflow I tried did not work.  And the workflow from above that does work requires that the user work directly within a clone of qmlweb.

Is it that use in other projects simply hasn't been addressed yet?

**Comment by Plaristote** [comment]
Well, it's far from being ready for use in web sites, IMHO. We have several ways of using the lib. If you plan on using it in a website, then the best solution is probably to use one of the release package (https://github.com/qmlweb/qmlweb/releases). They all come with the lib folder pre-built.

If you want to easily manage the qmlweb version you're using, you can also use `bower`. I've just tested it, `bower install qmlweb` will provide you with everything you need to use it in a website.

I still don't see how the Readme.md is wrong though. It proposes several solutions (including bower, as well as the github releases). And the cloning solution does mention that we need to run `npm run build`.

**Comment by ChALkeR** [comment]
@JoshuaKolden 
> So, we are in agreement, the Readme.md is wrong and the above steps require `npm run build`, yes?

No, Readme.md seems correct to me. `npm run build` is required _only when not using an already bundled release_, which is exactly what is written in the Readme.

**Comment by JoshuaKolden** [comment]
@Plaristote Well it's wrong in that it doesn't answer the question how does one use this project. The html example is different from what you provided, and qrc.js is not even mentioned.  Though I haven't checked if those issues make a difference yet.   The cloning solution does not get one to the point that they have a web page with a QML object on it, it stops at how to build but doesn't even explain what one is expecting to get as a result of that build.  At the risk of beating a dead horse "**how to use**" it is the question at hand.  It is also the the question of issue #240, and it is the title of that section of the Readme.  

In fact, this question "how does one use qmlweb" has been asked multiple times, by me and others.  I'm not alone in thinking this is poorly documented, yet the response to that seemingly innocent question has always been oddly tangential, and borderline hostile.  Which isn't to say that bothers me, but it will bother some.  I'm happy to help get to the bottom of this so we can close the gap, and that is what I'm attempting to do that here.

The readme suggests "See gulp-qmlweb package." which is were I come from originally regarding this issue (see the end of [issue #6][1] in that project).  What I'm understanding from the above is that that approach isn't expect to work (yet) outside of qmlweb.  Though  "qulp-" plugins are general assumed to deliver everything that is needed to use the given functionality (in a different project).

I don't spend a lot of time in Node.js so I don't know the ecosystem very well or idiomatic package management, so I think a readme with more clear "up and running" instructions would be an improvement. I would be happy to do it, but I'm clearly not the best candidate since I can't get anything to work consistently.  Though I will be happy to if I figure it out before anyone else updates the Readme.

---

@ChALkeR 
*"No, Readme.md seems correct to me. npm run build is required only when not using an already bundled release, which is exactly what is written in the Readme."*

**A)** No it isn't. This is what is written in the Readme (exactly):

```
## How to use
#### Add the library to your web page
Using one of the methods below, install the qmlweb JavaScript library:

- [npm](https://www.npmjs.com/package/qmlweb):

  npm install qmlweb

- [Bower](http://bower.io/search/?q=qmlweb):

  bower install qmlweb

- GitHub [releases](https://github.com/qmlweb/qmlweb/releases):

  tar -xzvf v0.0.4.tar.gz

- Manually using gulp (recommended if you cloned from git):

  npm install
  npm run build

Next, simply add `lib/qt.js` to the list of other JavaScript files in your app's HTML file:

...
```

*(I think you and I have very different understandings of the word "exactly", or we are looking at different "Readme.md" files.)*

The word 'bundled' doesn't even appear anywhere in the readme, and 'release' only in regards to obtaining the tar.gz, which is in the list of *alternatives* to the `npm install; npm run build;` commands. All under a section titled "How to use", not "How to build".  

**B)** What does "only when not using an already bundled release" mean?  What is a "release" and how is one supposed to use it?

[1]: https://github.com/qmlweb/gulp-qmlweb/issues/6

**Comment by ChALkeR** [comment]
@JoshuaKolden The Readme lists four different ways to get qmlweb. The first three don't require `npm run build` command, the last one does. That is consistent with what is written in the Readme.

**Comment by ChALkeR** [comment]
> What is a "release" and how is one supposed to use it?

Releases are everything that is tagged with a tag that fits under the following regex: `v[0-9]+.[0-9]+.[0-9]+`. They are avaiable though the tab that is named `[Releases]` here on GitHub ([link](https://github.com/qmlweb/qmlweb/releases)), through [npm](https://www.npmjs.com/package/qmlweb), and through Bower.

**Comment by JoshuaKolden** [comment]
..again for all those playing at home.

The question isn't "how do I get qmlweb?"

It's not "how do I build qmlweb?"

Can you guess what it is? 

What is the question?  Let's give it a second so the home audience has a chance to answer.  

The question is, wait for it... 

The question is:

How do I use qmlweb?

Bob, tell us what @ChALkeR's won.

"Well Josh, @ChALkeR's won an all expenses paid trip for 2 to Costa Rica, ((ding)) Oh no wait, I'm sorry.  I was just told by the judges that @ChALkeR's answer isn't correct!  Oh dear, I'm afraid @ChALkeR wins nothing." 

---

Ok, seriously though why is everyone dodging this question?

**Comment by ChALkeR** [comment]
@JoshuaKolden The Readme lists several ways to get qmlweb without building it. You copy-pasted them yourself here. Here are those reasons:

```md
Using one of the methods below, install the qmlweb JavaScript library:

- [npm](https://www.npmjs.com/package/qmlweb):

  npm install qmlweb

- [Bower](http://bower.io/search/?q=qmlweb):

  bower install qmlweb

- GitHub [releases](https://github.com/qmlweb/qmlweb/releases):

  tar -xzvf v0.0.4.tar.gz
```

Only the last one is related to manually building it:
```md
- Manually using gulp (recommended if you cloned from git):

  npm install
  npm run build
```

**Comment by ChALkeR** [comment]
@JoshuaKolden Ah.

For the «how do I use qmlweb», you could try following the example at the Readme in this repo, without gulp-qmlweb first. That should work. Once you get that working, you can offload parsing to the server using gulp-qmlweb.

**Comment by JoshuaKolden** [comment]
I get the impression people aren't reading the actual thread, but nevertheless I think I have the workflow given the one that almost worked from @Plaristote above.  I'll check the <script> tag in both place in the HTML just to be sure that isn't a factor.  I think I have a test grid now so I can lay out the different ways I've tried, and those suggested above.  I'll put this all in a separate project and link to it.

What still doesn't seem to work, but will be part of what I test is just including the lib/* folder in a web project and using the above html.  It works in the trivial case of a "Hello, world" qml file, but the `import` commands don't seem to work in that context.

Thanks. 

I'm working on this now so I'll post my findings in a bit.

**Comment by ChALkeR** [comment]
@JoshuaKolden Once again — I wasn't able to reproduce the issue that you are observing. Could you please upload a failing example to some webserver?

**Comment by Plaristote** [comment]
@JoshuaKolden 
I'm going to slowly step away from this (until we have an archive that allows us to reproduce the `import` issue). Just wanted to say that it's my bad: I didn't have to write a new html file, the one from the Readme works as well...

And `qrc` is an optional thing, it will be used if you're using `gulp-qmlweb`, so it should be explained in the `gulp-qmlweb` readme. I think it would only be more confusing to see it in the `qmlweb` readme, since people who won't use `gulp-qmlweb` won't even have a `qrc.js` file.

**Comment by JoshuaKolden** [comment]
@stephenmdangelo I can now confirm the issue I'm seeing is that `import QtQuick.Controls 1.4` must be in *all* qml files. All imports must be in all qml files.  I have not yet tested  #239, but will shortly.

**Comment by JoshuaKolden** [comment]
https://github.com/JoshuaKolden/qmlwebtest

**Comment by JoshuaKolden** [comment]
@stephenmdangelo I finally got around to testing #239, but it doesn't appear to be working.  You can take a look at the `test_pr239` folder in the above qmlwebtest link.

**Comment by stephenmdangelo** [comment]
@JoshuaKolden ah, due to some changes in master after I originally made this PR, it now depends on #274 as well. I had rebased my local branch on top of it, but I guess I didn't push it. I will do that tomorrow, or you can cherry-pick the two commits from #239 on top of #274 if you feel like testing before then.

**Comment by JoshuaKolden** [comment]
@stephenmdangelo Yeah I thought there might be something like that.  I did it again, and it works!

Excellent work @stephenmdangelo

**Comment by JoshuaKolden** [comment]
So it was a known bug all along!

Wow, wasn't that like pulling teeth!?  That's about as much resistance to resolving an issue as I've ever seen on github.  Oh well.  Good exercise in due diligence I suppose.

**Comment by stephenmdangelo** [comment]
@JoshuaKolden, it appears that #239 was up-to-date all along (e.g. it's already re-based on the latest #274, e9c2fda4e316b92d2f1dbd7f45a95247b3c386b8), so I'm not sure how test_pr239 initially ended up with the outdated version of the code.

**Comment by JoshuaKolden** [comment]
@stephenmdangelo, No doubt I pulled it incorrectly the first time than.

**Comment by Plaristote** [comment]
Sounds like the issue was solved, I'm closing this.

**Comment by stephenmdangelo** [comment]
@Plaristote This is not solved. It is solved by #239, but that has yet to be merged.

**Comment by stephenmdangelo** [comment]
Should be fixed now as #239 has landed.
